net/sched/sch_gkprio.c

```diff
@@ -0,0 +1,306 @@
+
+	if (prio > q->highest_prio)
+		q->highest_prio = prio;
+
+	return NET_XMIT_SUCCESS;
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
+}
+
+static struct sk_buff *gkprio_dequeue(struct Qdisc *sch)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct sk_buff_head *hpq = &q->qdiscs[q->highest_prio];
+	struct sk_buff *skb = __skb_dequeue(hpq);
+
+	if (unlikely(!skb))
+		return NULL;
+
+	sch->q.qlen--;
+	qdisc_qstats_backlog_dec(sch, skb);
+	qdisc_bstats_update(sch, skb);
+
+	/* Update highest priority field. */
+	if (skb_queue_empty(hpq))
+		if (q->lowest_prio == q->highest_prio) {
+			q->highest_prio = 0;
+			q->lowest_prio = GKPRIO_MAX_PRIORITY - 1;
+		}
+		else
+			q->highest_prio = calc_new_high_prio(q);
+
+	return skb;
+}
+
+static int gkprio_init(struct Qdisc *sch, struct nlattr *opt)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	int prio;
+
+	if (!opt) {
+		q->max_limit = qdisc_dev(sch)->tx_queue_len;
+		q->noip_dfltp = 0;
+	} else {
+		struct tc_gkprio_qopt *ctl = nla_data(opt);
+
+		if (nla_len(opt) < sizeof(*ctl))
+			return -EINVAL;
+
+		if (ctl->limit == (typeof(ctl->limit)) - 1)
+			q->max_limit = qdisc_dev(sch)->tx_queue_len;
+		else if (ctl->limit > qdisc_dev(sch)->tx_queue_len) 
+			return -EINVAL;
+		else
+			q->max_limit = ctl->limit;
+		if (!q->max_limit)
+			return -EINVAL;		
+		if (ctl->dfltp == (typeof(ctl->dfltp)) - 1) 
+			q->noip_dfltp = 0;
+		else if (ctl->dfltp >= GKPRIO_MAX_PRIORITY)
+			return -EINVAL;
+		else
+			q->noip_dfltp = ctl->dfltp;
+	}
+
+	/* Initialise 64 queues, one for each possible priority. */
+	for (prio = 0; prio < GKPRIO_MAX_PRIORITY; prio++)
+		__skb_queue_head_init(&q->qdiscs[prio]);
+
+	q->highest_prio = 0;
+	q->lowest_prio = GKPRIO_MAX_PRIORITY - 1;
+	return 0;
+}
+
+static int gkprio_dump(struct Qdisc *sch, struct sk_buff *skb)
+{
+  	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct tc_gkprio_qopt opt;
+
+	opt.limit = q->max_limit;
+	opt.dfltp = q->noip_dfltp;
+	opt.highest_prio = q->highest_prio;
+	opt.lowest_prio = q->lowest_prio;
+
+	if (nla_put(skb, TCA_OPTIONS, sizeof(opt), &opt))
+		return -1;
+
+	return skb->len;
+}
+
+static void gkprio_reset(struct Qdisc *sch)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	int prio;
+
+	sch->qstats.backlog = 0;
+	sch->q.qlen = 0;
+
+	for (prio = 0; prio < GKPRIO_MAX_PRIORITY; prio++)
+		skb_queue_purge(&q->qdiscs[prio]);
+	q->highest_prio = 0;
+	q->lowest_prio = GKPRIO_MAX_PRIORITY - 1;
+}
+
+static void gkprio_destroy(struct Qdisc *sch)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	int prio;
+
+	for (prio = 0; prio < GKPRIO_MAX_PRIORITY; prio++)
+		skb_queue_purge(&q->qdiscs[prio]);
+}
+
+struct Qdisc_ops gkprio_qdisc_ops __read_mostly = {
+	.id		=	"gkprio",
+	.priv_size	=	sizeof(struct gkprio_sched_data),
+	.enqueue	=	gkprio_enqueue,
+	.dequeue	=	gkprio_dequeue,
+	.peek		=	qdisc_peek_dequeued,
+	.init		=	gkprio_init,
+	.reset		=	gkprio_reset,
+	.change		=	gkprio_init,
```

Comment:
`gkprio_init()` updates the parameters *and* initializes the queues, whereas `.change` expects to only update the parameters.The problem can be easy addressed breaking up `gkprio_init()` into two functions.